zerver/tornado/event_queue.py

```diff
@@ -1637,6 +1637,24 @@ def process_stream_typing_notification_event(event: Mapping[str, Any], users: It
             client.add_event(event_copy)
 
 
+def process_mark_message_unread_event(event: Mapping[str, Any], users: Iterable[int]) -> None:
+    for user_profile_id in users:
+        for client in get_client_descriptors_for_user(user_profile_id):
+            if not client.accepts_event(event):
+                continue
+
+            event_copy = copy.deepcopy(dict(event))
+            for message_id, message_detail in event_copy["message_details"].items():
+                if (
+                    message_detail["type"] == "stream"
+                    and message_detail.get("topic") == ""
+                    and not client.empty_topic_name
+                ):
+                    event_copy["message_details"][message_id]["topic"] = "general chat"
+
+            client.add_event(event_copy)
```

Comment:
This can probably be optimized to skip the copying step in the (default for the web app) case where `client.empty_topic_name`. I think generally we do want to write this event_queue.py code that way, since it is performa-sensitive code, though this particular event is rare enough that it likely doesn't matter.

Also, rather than hardcoding "general chat" here, we probably want to be using a global, like `Message.EMPTY_TOPIC_FALLBACK_NAME`, to avoid having lots of hardcoded references to that topic name..

Reply:
> Also, rather than hardcoding "general chat" here, we probably want to be using a global, like Message.EMPTY_TOPIC_FALLBACK_NAME

Done

> This can probably be optimized to skip the copying step in the case where client.empty_topic_name

Done

Reply:
> This can probably be optimized to skip the copying step in the case where client.empty_topic_name

Based on this feedback, did some optimization in `process_user_topic_event` and `process_stream_typing_notification_event`.

Reply:
I posted some ideas for how to do that more nicely in another comment; this code can probably do the same approach of doing the `deepcopy` once up front.

Reply:
Made the changes to perform deepcopy once.